File: src/lib/golingvu/base_it.go
package golingvu

// BaseIt extends BaseAction for Describe-It pattern
type BaseIt struct {
	*BaseAction
	ItCB interface{}
}

// NewBaseIt creates a new BaseIt instance
func NewBaseIt(name string, itCB interface{}) *BaseIt {
	baseAction := NewBaseAction(name, itCB)
	return &BaseIt{
		BaseAction: baseAction,
		ItCB:       itCB,
	}
}

// PerformAction implements BaseAction's abstract method
func (bi *BaseIt) PerformAction(
	store interface{},
	actionCB interface{},
	testResource interface{},
	artifactory func(string, interface{}),
) (interface{}, error) {
	// This should be implemented by the adapter
	// It can perform both mutations and assertions
	return store, nil
}

// PerformIt is an alias for PerformAction for Describe-It pattern
func (bi *BaseIt) PerformIt(
	store interface{},
	itCB interface{},
	testResource interface{},
	artifactory func(string, interface{}),
) (interface{}, error) {
	return bi.PerformAction(store, itCB, testResource, artifactory)
}

// Test executes the It with artifactory
func (bi *BaseIt) Test(
	store interface{},
	testResourceConfiguration ITTestResourceConfiguration,
	artifactory func(string, interface{}),
) (interface{}, error) {
	return bi.PerformAction(store, bi.ItCB, testResourceConfiguration, artifactory)
}

File: src/lib/rusto/src/base_it.rs
use crate::base_action::BaseAction;
use crate::types::IbddInAny;
use std::collections::HashMap;

/// BaseIt extends BaseAction for Describe-It pattern.
/// Its can mix mutations and assertions, unlike BDD's When which only does mutations.
pub struct BaseIt<I: IbddInAny> {
    inner: BaseAction<I>,
}

impl<I: IbddInAny> BaseIt<I> {
    pub fn new(name: String, it_cb: Box<dyn Fn(I::Iselection) -> I::Then>) -> Self {
        Self {
            inner: BaseAction::new(name, it_cb),
        }
    }
    
    pub fn add_artifact(&mut self, path: String) {
        self.inner.add_artifact(path);
    }
    
    pub fn to_obj(&self) -> HashMap<String, Box<dyn std::any::Any>> {
        self.inner.to_obj()
    }
    
    // Alias test to it for Describe-It pattern
    pub async fn it(
        &mut self,
        store: I::Istore,
        test_resource: &dyn std::any::Any,
    ) -> Result<I::Istore, String> {
        self.inner.test(store, test_resource).await
    }
    
    // Perform it with artifactory
    pub async fn perform_it(
        &mut self,
        store: I::Istore,
        test_resource: &dyn std::any::Any,
        _artifactory: Option<&dyn std::any::Any>,
    ) -> Result<I::Istore, String> {
        // In a real implementation, artifactory would be used
        self.inner.test(store, test_resource).await
    }
}
